testing/src/scenario/context.py

```diff
@@ -688,26 +704,172 @@ def __init__(
         self.capture_framework_events = capture_framework_events
 
         # streaming side effects from running an event
-        self.juju_log: list[JujuLogLine] = []
-        self.app_status_history: list[_EntityStatus] = []
-        self.unit_status_history: list[_EntityStatus] = []
-        self.exec_history: dict[str, list[ExecArgs]] = {}
-        self.workload_version_history: list[str] = []
-        self.removed_secret_revisions: list[int] = []
-        self.emitted_events: list[ops.EventBase] = []
-        self.requested_storages: dict[str, int] = {}
-        self.trace_data = []
+        self._juju_log: list[JujuLogLine] = []
+        self._app_status_history: list[_EntityStatus] = []
+        self._unit_status_history: list[_EntityStatus] = []
+        self._exec_history: dict[str, list[ExecArgs]] = {}
+        self._workload_version_history: list[str] = []
+        self._removed_secret_revisions: list[int] = []
+        self._emitted_events: list[ops.EventBase] = []
+        self._requested_storages: dict[str, int] = {}
+        self._trace_data = []
 
         # set by Runtime.exec() in self._run()
         self._output_state: State | None = None
+        # set if we're being used to create a context manager (None otherwise)
+        self._manager: Manager[CharmType] | None = None
 
         # operations (and embedded tasks) from running actions
-        self.action_logs: list[str] = []
-        self.action_results: dict[str, Any] | None = None
+        self._action_logs: list[str] = []
+        self._action_results: dict[str, Any] | None = None
         self._action_failure_message: str | None = None
 
         self.on = CharmEvents()
 
+    def _deprecated_side_effect_access(self, attr: str):
+        warnings.warn(
+            f"Accessing '{attr}' on the Context object is deprecated. Use the Context as a "
+            f"context manager instead, with 'mgr.{attr}`",
+            DeprecationWarning,
+            stacklevel=3,
+        )
+        return getattr(self, f'_{attr}')
+
+    @property
+    def juju_log(self) -> list[JujuLogLine]:
+        """A record of what the charm has sent to juju-log"""
+        return self._deprecated_side_effect_access('juju_log')
+
+    @property
+    def app_status_history(self) -> list[_EntityStatus]:
+        """A record of the app statuses the charm has set.
+
+        Assert that the charm has followed the expected path by checking the app
```

Comment:
To avoid the doc duplication, should we trim all these docstrings down to just the one-line version, with a second paragraph that says "This field is deprecated, use <link to new thing> instead."